pkg/controller.v2/controller_pod.go

```diff
@@ -0,0 +1,358 @@
+			}
+
+			// Generate TF_CONFIG JSON string.
+			tfConfigStr := genTFConfigJSONStr(tfjob, rt, index)
+
+			if tfConfigStr == "" {
+				return nil
+			}
+			// Add TF_CONFIG environment variable.
+			for i, _ := range pTemplate.Spec.Containers {
+				if len(pTemplate.Spec.Containers[i].Env) == 0 {
+					pTemplate.Spec.Containers[i].Env = make([]v1.EnvVar, 0)
+				}
+				pTemplate.Spec.Containers[i].Env = append(pTemplate.Spec.Containers[i].Env, v1.EnvVar{
+					Name:  "TF_CONFIG",
+					Value: tfConfigStr,
+				})
+			}
+
+			// Set restart policy
+			if spec.RestartPolicy != "ExitCode" {
+				pTemplate.Spec.RestartPolicy = v1.RestartPolicy(spec.RestartPolicy)
+			}
+
+			err := tc.podControl.CreatePodsWithControllerRef(tfjob.Namespace, pTemplate, tfjob, controllerRef)
+			if err != nil && errors.IsTimeout(err) {
+				// Pod is created but its initialization has timed out.
+				// If the initialization is successful eventually, the
+				// controller will observe the creation via the informer.
+				// If the initialization fails, or if the pod keeps
+				// uninitialized for a long time, the informer will not
+				// receive any update, and the controller will create a new
+				// pod when the expectation expires.
+				return nil
+			} else if err != nil {
+				return err
+			}
+		}
+	} else if diff > 0 {
+		// TODO(CPH): Need to delete pods.
+		log.Infof("need to delete pod but it is not implemented yet")
+	}
+
+	if tfjob.Status.TFReplicaStatuses == nil {
+		tfjob.Status.TFReplicaStatuses = make(map[tfv1alpha2.TFReplicaType]*tfv1alpha2.TFReplicaStatus)
+	}
+
+	if _, ok := tfjob.Status.TFReplicaStatuses[rtype]; !ok {
+		tfjob.Status.TFReplicaStatuses[rtype] = &tfv1alpha2.TFReplicaStatus{}
+	}
+
+	// Update the active status since we have created -diff pods during the loop.
+	tfjob.Status.TFReplicaStatuses[rtype].Active = expected
+	tfjob.Status.TFReplicaStatuses[rtype].Succeeded = succeeded
+	tfjob.Status.TFReplicaStatuses[rtype].Failed = failed
+	return nil
+}
+
+// getDiffPodIndexes checks and gets diff indexes from desired and current.
+func getDiffPodIndexes(activePods []*v1.Pod, replicas int32) []string {
+	desiredIndexes := make(map[string]string)
+
+	for i := int32(0); i < replicas; i++ {
+		desiredIndexes[fmt.Sprintf("%d", i)] = noHit
+	}
+
+	for _, pod := range activePods {
+		if _, ok := pod.Labels[tfReplicaIndexLabel]; !ok {
+			continue
+		}
+
+		index := pod.Labels[tfReplicaIndexLabel]
+
+		if _, ok := desiredIndexes[index]; ok {
+			desiredIndexes[index] = hit
+		}
+	}
+
+	diffIndexes := []string{}
+	for index, hit := range desiredIndexes {
+		if hit == noHit {
+			diffIndexes = append(diffIndexes, index)
+		}
+	}
+
+	return diffIndexes
+}
+
+// getPodsForTFJob returns the set of pods that this tfjob should manage.
+// It also reconciles ControllerRef by adopting/orphaning.
+// Note that the returned Pods are pointers into the cache.
+func (tc *TFJobController) getPodsForTFJob(tfjob *tfv1alpha2.TFJob) ([]*v1.Pod, error) {
```

Comment:
Yeah I think we should change the name to `getAndClaimPodsForTFJob`. And I think it is better to remove the logic of claiming， since there is no need to claim the pods every time. I can open an issue for it.

WDYT @ScorpioCPH

Reply:
https://github.com/kubeflow/tf-operator/issues/533

Reply:
this function name follows the other controller style, like this https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L401.

Reply:
SGTM